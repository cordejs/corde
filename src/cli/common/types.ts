import { Command } from "commander";

export type ParamsFrom = "args" | "options" | "both";

export interface ICommandOptions {
  flags: string;
  description?: string;
  defaultValue?: string;
}

export interface ICliCommandConstructor {
  name?: string;
  program: Command;
  paramsFrom?: ParamsFrom;
}

export interface ICliCommand {
  readonly paramsFrom?: ParamsFrom;
  readonly command: Command;
  handler(...args: any[]): Promise<void> | void;
}

export interface ICliCommandConstructable {
  new (command: Command): ICliCommand;
}
