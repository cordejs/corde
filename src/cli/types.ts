export interface ICommand {
  action(...args: any[]): Promise<void> | void;
}
