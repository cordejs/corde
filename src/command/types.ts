import { ICordeBot } from "../types";

export interface ICreateMatcherParam {
  matcher?: string;
  isNot?: boolean;
  commandName?: string | boolean | number;
  channelId?: string;
  guildId?: string;
  isDebug?: boolean;
  isCascade?: boolean;
  cordeBot?: ICordeBot;
  trace?: string;
  utility?: boolean;
  mustSendCommand?: boolean;
}
