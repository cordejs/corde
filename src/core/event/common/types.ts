import { Client, ClientEvents } from "discord.js";

export type EventNames = keyof ClientEvents;

export type Spread<T extends EventNames> = ClientEvents[T][1] extends undefined
  ? ClientEvents[T] extends (infer U)[]
    ? U
    : never
  : ClientEvents[T];

export interface IDiscordEvent<TEvent extends EventNames, TFilter> {
  canListen(): boolean;
  getIntents(): string[];
  on(fn: (...args: ClientEvents[TEvent]) => void): this;
  once(filter?: TFilter): Promise<Spread<TEvent>>;
}

export interface IDiscordEventConstructable<TEvent extends EventNames, TFilter> {
  new (client: Client): IDiscordEvent<TEvent, TFilter>;
}

export type IntentType =
  | "GUILDS"
  | "GUILD_MEMBERS"
  | "GUILD_BANS"
  | "GUILD_EMOJIS_AND_STICKERS"
  | "GUILD_INTEGRATIONS"
  | "GUILD_WEBHOOKS"
  | "GUILD_INVITES"
  | "GUILD_VOICE_STATES"
  | "GUILD_PRESENCES"
  | "GUILD_MESSAGES"
  | "GUILD_MESSAGE_REACTIONS"
  | "GUILD_MESSAGE_TYPING"
  | "DIRECT_MESSAGES"
  | "DIRECT_MESSAGE_REACTIONS"
  | "DIRECT_MESSAGE_TYPING"
  | "GUILD_SCHEDULED_EVENTS";

export type EventClass = {
  [k in IntentType]: {
    bitfield: number;
    events: EventNames[];
  };
};
