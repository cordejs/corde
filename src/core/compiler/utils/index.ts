export * from "./import-js-in-directory";
export * from "./was-file-modified";
