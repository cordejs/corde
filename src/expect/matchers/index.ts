export * from "./toBeArray";
export * from "./toBeBigInt";
export * from "./toBeBoolean";
export * from "./toBeDate";
export * from "./toBeDefined";
export * from "./toBeEmptyArray";
export * from "./toBeEmptyString";
export * from "./toBeFunction";
export * from "./toBeGreaterOrEqualThan";
export * from "./toBeGreaterThan";
export * from "./toBeLessOrEqualThan";
export * from "./toBeLessThan";
export * from "./toBeNaN";
export * from "./toBeNothing";
export * from "./toBeNull";
export * from "./toBeUndefined";
export * from "./toBeNumber";
export * from "./toBePrimitive";
export * from "./toBeString";
export * from "./toBeSymbol";
export * from "./toBeValidDate";
export * from "./toLength";
export * from "./toBeInstanceOf";
export * from "./toBeNegativeInfinity";
export * from "./toBePositiveInfinity";
export * from "./toBe";
