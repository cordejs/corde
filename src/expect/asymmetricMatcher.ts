import { typeOf } from "../utils/typeOf";

export class AsymmetricMatcher {
  private _classType: any[];

  constructor(...classType: any[]) {
    this._classType = classType;
  }

  isSpecified() {
    return this._classType.length > 0;
  }

  toString() {
    if (this.isSpecified()) {
      const names = this._classType.map((type) => this.getTypeName(type));
      return `any(${names.join(", ")})`;
    }
    return "any";
  }

  getTypeName(classType: any) {
    if (classType === Number) {
      return "Number";
    }

    if (classType === Date) {
      return "Date";
    }

    if (classType === BigInt) {
      return "Bigint";
    }

    if (classType === String) {
      return "String";
    }

    if (classType === Boolean) {
      return "Boolean";
    }

    if (classType === Array) {
      return "Array";
    }

    if (classType === Object) {
      return "Object";
    }

    if (classType === Function) {
      return "Function";
    }

    if (classType === Symbol) {
      return "Symbol";
    }

    if (typeof classType === "function") {
      return classType.name;
    }

    return classType;
  }

  getTypes() {
    return this._classType;
  }

  matchType(...classType: any[]) {
    return (
      this._classType === undefined ||
      this._classType === null ||
      this._classType.length === 0 ||
      classType.length === 0 ||
      this._classType.some((type) => classType.includes(type))
    );
  }

  matchValue(value: any) {
    return (
      this.isSpecified() ||
      this._classType.every((classType) => this._isValueOfType(classType, value))
    );
  }

  private _isValueOfType(classType: any, value: any) {
    if (classType === Number) {
      return typeOf(value) === "number" || value instanceof Number;
    }

    if (classType === BigInt) {
      return typeOf(value) === "bigint" || value instanceof BigInt;
    }

    if (classType === String) {
      return typeOf(value) === "string" || value instanceof String;
    }

    if (classType === Boolean) {
      return typeOf(value) === "boolean" || value instanceof Boolean;
    }

    if (classType === Array) {
      return typeOf(value) === "array";
    }

    if (classType === Object) {
      return typeOf(value) === "object";
    }

    if (classType === Function) {
      return typeOf(value) === "function" || value instanceof Function;
    }

    if (classType === Symbol) {
      return typeOf(value) === "symbol";
    }

    if (typeOf(value) === "object") {
      return value instanceof classType;
    }

    return false;
  }
}

export function any(...classType: any[]) {
  return new AsymmetricMatcher(...classType);
}
