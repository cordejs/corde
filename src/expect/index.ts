/* eslint-disable no-console */
import chalk from "chalk";
import { runtime } from "../common/runtime";
import { ITestProps, ITestReport } from "../types";
import { corde } from "../types/globals";
import { getStackTrace, typeOf } from "../utils";
import { any } from "./asymmetricMatcher";
import * as matchers from "./matchers";

interface IMatcher {
  (props: ITestProps, ...args: any[]): ITestReport;
}

type KeyOfMatcher = keyof typeof matchers;

function pickFn(name: KeyOfMatcher) {
  return matchers[name] as any as IMatcher;
}

function createMatcherFn(matcher: string, isNot: boolean, expected: any, isDebug: boolean) {
  return (...args: any[]): ITestReport | void => {
    // If someone pass expect.any, we must invoke it to return
    // the Any matcher.

    if (expected === any) {
      args = [expected(), ...args];
    } else {
      args = [expected, ...args];
    }

    args = args.map((arg) => {
      if (arg === any) {
        return arg();
      }
      return arg;
    });

    const trace = getStackTrace(undefined, true, matcher);
    try {
      const matcherFn = pickFn(matcher as KeyOfMatcher);

      const props: ITestProps = {
        isNot,
        expectedColorFn: chalk.bold,
        receivedColorFn: chalk.bold,
        createHint: (...paramsName: string[]) => {
          return `expect(${chalk.green("expected")}).${isNot ? "not." : ""}${matcher}(${paramsName
            .map((param) => chalk.red(param))
            .join(",")})`;
        },
        formatValue: (value: any) => {
          if (typeof value === "symbol") {
            return value.toString();
          }

          if (typeOf(value) === "object") {
            if (Object.keys(value).length === 0) {
              return "{}";
            }

            return "{ ... }";
          }

          if (
            typeof value === "string" &&
            (value.length === 0 || (!value.startsWith("'") && !value.endsWith("'")))
          ) {
            return `'${value}'`;
          }

          if (Array.isArray(value)) {
            if (value.length > 0) {
              return "[...]";
            }
            return "[]";
          }

          return value;
        },
      };

      const report = matcherFn.bind(props, ...args)();
      if (!report.pass) {
        report.trace = trace;
      }

      runtime.internalEvents.emit("test_end", report);

      if (isDebug) {
        return report;
      }
    } catch (error) {
      const failedReport: ITestReport = {
        pass: false,
        message: handleError(error),
        trace: trace,
      };
      runtime.internalEvents.emit("test_end", failedReport);
      if (isDebug) {
        return failedReport;
      }
    }
  };
}

function handleError(error: any) {
  if (error instanceof Error) {
    return error.message;
  }
  return error;
}

function createLocalExpect(isDebug: boolean) {
  let localExpect: any = {};

  localExpect = (expected: any) => {
    const _expect: any = {};
    _expect.not = {};
    Object.getOwnPropertyNames(matchers).forEach((matcher) => {
      _expect[matcher] = createMatcherFn(matcher, false, expected, isDebug);
      _expect.not[matcher] = createMatcherFn(matcher, true, expected, isDebug);
    });
    return _expect;
  };

  localExpect.any = any;

  return localExpect;
}

export const expect = createLocalExpect(false) as corde.IExpect;

type DebugExpectType<T> = {
  [P in keyof T]: T[P] extends (...args: any[]) => any
    ? (...params: Parameters<T[P]>) => { pass: boolean; message: string }
    : DebugExpectType<T[P]>;
};

export interface IDebugExpect {
  <T extends any>(value: T): DebugExpectType<corde.IMatchersWithNot<T>>;
  any(...classType: any[]): any;
}

/**
 * Testing object for corde
 * This is not used in production
 *
 * @internal
 */
export const cordeExpect = createLocalExpect(true) as IDebugExpect;
