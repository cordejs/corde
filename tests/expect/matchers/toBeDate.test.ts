import chalk from "chalk";
import { any } from "../../../src/expect/asymmetricMatcher";
import { toBeDate } from "../../../src/expect/matchers";
import { ITestProps } from "../../../src/types";
import { buildReportMessage, typeOf } from "../../../src/utils";
import { removeANSIColorStyle } from "../../testHelper";

function getMessage(expected: any, isNot: boolean) {
  const isNotText = isNot ? " not" : "";
  return buildReportMessage(
    `expected type${isNotText} to be ${chalk.green("Date")}.\n`,
    `received: '${chalk.red(typeOf(expected))}'`,
  );
}

const TEST_CASES = [[null], [1], [undefined], ["aa"], [{}], [Symbol.for("")]];

describe("testing toBeDefined", () => {
  it("should return true for a Date value", () => {
    expect(toBeDate({ isNot: false }, new Date())).toEqual({ pass: true, message: "" });
  });

  it("should return true for asymmetricMatcher", () => {
    expect(toBeDate({ isNot: false }, any())).toEqual({ pass: true, message: "" });
    expect(toBeDate({ isNot: false }, any(Date))).toEqual({ pass: true, message: "" });
  });

  it("should return false for asymmetricMatcher of any value that is not Date", () => {
    expect(toBeDate({ isNot: false }, any(Number))).toEqual({
      pass: false,
      message: getMessage(any(Number), false),
    });
  });

  it("should return false for Date with isNot true", () => {
    const props: ITestProps = { isNot: true };
    const response = toBeDate(props, new Date());
    expect(response).toEqual({
      pass: false,
      message: getMessage(new Date(), props.isNot),
    });
    response.message = removeANSIColorStyle(response.message);
    expect(response).toMatchSnapshot();
  });

  // @ts-ignore
  it.each(TEST_CASES)("should return false for %s", (expected) => {
    const props: ITestProps = { isNot: false };
    const response = toBeDate(props, expected);
    expect(response).toEqual({
      pass: false,
      message: getMessage(expected, props.isNot),
    });
    response.message = removeANSIColorStyle(response.message);
    expect(response).toMatchSnapshot();
  });

  it.each(TEST_CASES)(
    "should return true for %s, when isNot is true",
    // @ts-ignore
    (expected) => {
      const props: ITestProps = { isNot: true };
      expect(toBeDate(props, expected)).toEqual({
        pass: true,
        message: "",
      });
    },
  );
});
