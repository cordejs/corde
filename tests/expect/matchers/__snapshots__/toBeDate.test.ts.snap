// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing toBeDefined should return false for {} 1`] = `
Object {
  "message": "expect([32mexpected[39m).toBeDate()

[1mexpected[22m should be a [32mdate[39m.
got: [31mobject[39m",
  "pass": false,
}
`;

exports[`testing toBeDefined should return false for 1 1`] = `
Object {
  "message": "expect([32mexpected[39m).toBeDate()

[1mexpected[22m should be a [32mdate[39m.
got: [31mnumber[39m",
  "pass": false,
}
`;

exports[`testing toBeDefined should return false for Date with isNot true 1`] = `
Object {
  "message": "expect([32mexpected[39m).not.toBeDate()

[1mexpected[22m should not be a [32mdate[39m.
got: [31mobject[39m",
  "pass": false,
}
`;

exports[`testing toBeDefined should return false for Symbol() 1`] = `
Object {
  "message": "expect([32mexpected[39m).toBeDate()

[1mexpected[22m should be a [32mdate[39m.
got: [31msymbol[39m",
  "pass": false,
}
`;

exports[`testing toBeDefined should return false for aa 1`] = `
Object {
  "message": "expect([32mexpected[39m).toBeDate()

[1mexpected[22m should be a [32mdate[39m.
got: [31mstring[39m",
  "pass": false,
}
`;

exports[`testing toBeDefined should return false for asymmetricMatcher of any value that is not Date 1`] = `
Object {
  "message": "expect([32mexpected[39m).toBeDate()

[1mexpected[22m should be a [32mdate[39m.
got: [31many(Number)[39m",
  "pass": false,
}
`;

exports[`testing toBeDefined should return false for null 1`] = `
Object {
  "message": "expect([32mexpected[39m).toBeDate()

[1mexpected[22m should be a [32mdate[39m.
got: [31mnull[39m",
  "pass": false,
}
`;

exports[`testing toBeDefined should return false for undefined 1`] = `
Object {
  "message": "expect([32mexpected[39m).toBeDate()

[1mexpected[22m should be a [32mdate[39m.
got: [31mundefined[39m",
  "pass": false,
}
`;
